File: apps/patter-voice-sdk/libraries/typescript/src/providers/telnyx-stt.ts
/**
 * Telnyx Speech-to-Text adapter (WebSocket streaming).
 *
 * Bridges the Telnyx `/v2/speech-to-text/transcription` WebSocket API to the
 * Patter SDK pipeline-mode STT interface. Implemented in TypeScript
 * (`ws` + `Buffer`) with a callback-based interface matching the other
 * Patter STT providers (Deepgram, Whisper).
 */

import WebSocket from 'ws';
import { getLogger } from '../logger';

/** Patter-normalised transcript event emitted by {@link TelnyxSTT}. */
export interface Transcript {
  readonly text: string;
  readonly isFinal: boolean;
  readonly confidence: number;
}

type TranscriptCallback = (transcript: Transcript) => void;

/** Backing transcription engine accepted by Telnyx STT. */
export type TelnyxTranscriptionEngine = 'telnyx' | 'google' | 'deepgram' | 'azure';

/** Common PCM sample rates accepted by Telnyx STT. */
export const TelnyxSTTSampleRate = {
  HZ_8000: 8000,
  HZ_16000: 16000,
  HZ_24000: 24000,
} as const;
/** Union of {@link TelnyxSTTSampleRate} integer values. */
export type TelnyxSTTSampleRate =
  (typeof TelnyxSTTSampleRate)[keyof typeof TelnyxSTTSampleRate];

/** Input audio formats accepted by Telnyx STT. */
export const TelnyxSTTInputFormat = {
  WAV: 'wav',
} as const;
/** Union of {@link TelnyxSTTInputFormat} string values. */
export type TelnyxSTTInputFormat =
  (typeof TelnyxSTTInputFormat)[keyof typeof TelnyxSTTInputFormat];

const TELNYX_STT_WS_URL = 'wss://api.telnyx.com/v2/speech-to-text/transcription';
const DEFAULT_SAMPLE_RATE: TelnyxSTTSampleRate = TelnyxSTTSampleRate.HZ_16000;
const NUM_CHANNELS = 1;

/** Build a streaming WAV header with maximum possible data size. */
function createStreamingWavHeader(sampleRate: number, numChannels: number): Buffer {
  const bytesPerSample = 2;
  const byteRate = sampleRate * numChannels * bytesPerSample;
  const blockAlign = numChannels * bytesPerSample;
  const dataSize = 0x7fffffff;
  const fileSize = 36 + dataSize;

  const header = Buffer.alloc(44);
  header.write('RIFF', 0);
  header.writeUInt32LE(fileSize, 4);
  header.write('WAVE', 8);
  header.write('fmt ', 12);
  header.writeUInt32LE(16, 16);
  header.writeUInt16LE(1, 20); // PCM
  header.writeUInt16LE(numChannels, 22);
  header.writeUInt32LE(sampleRate, 24);
  header.writeUInt32LE(byteRate, 28);
  header.writeUInt16LE(blockAlign, 32);
  header.writeUInt16LE(16, 34); // bits per sample
  header.write('data', 36);
  header.writeUInt32LE(dataSize, 40);
  return header;
}

/** Streaming STT adapter for Telnyx's `/v2/speech-to-text` WebSocket. */
export class TelnyxSTT {
  /** Stable pricing/dashboard key — read by stream-handler/metrics. */
  static readonly providerKey = 'telnyx_stt';
  private ws: WebSocket | null = null;
  private callbacks: Set<TranscriptCallback> = new Set();
  private headerSent = false;

  /** Construction args replayed by clone(). */
  private readonly patterCtorArgs: unknown[];

  constructor(
    private readonly apiKey: string,
    private readonly language: string = 'en',
    private readonly transcriptionEngine: TelnyxTranscriptionEngine = 'telnyx',
    private readonly sampleRate: number = DEFAULT_SAMPLE_RATE,
    private readonly baseUrl: string = TELNYX_STT_WS_URL,
  ) {
    this.patterCtorArgs = [apiKey, language, transcriptionEngine, sampleRate, baseUrl];
  }

  /**
   * Fresh adapter built with this instance's construction arguments —
   * called per call by the stream handler so concurrent calls never share
   * connection state (sockets/queues; cross-call transcript bleed).
   */
  clone(): this {
    const ctor = this.constructor as new (...args: unknown[]) => this;
    return new ctor(...this.patterCtorArgs);
  }


  /** Open the streaming WebSocket and arm message handlers. */
  async connect(): Promise<void> {
    const params = new URLSearchParams({
      transcription_engine: this.transcriptionEngine,
      language: this.language,
      input_format: 'wav',
    });
    const url = `${this.baseUrl}?${params.toString()}`;

    this.ws = new WebSocket(url, {
      headers: { Authorization: `Bearer ${this.apiKey}` },
    });

    await new Promise<void>((resolve, reject) => {
      const timer = setTimeout(() => reject(new Error('Telnyx STT connect timeout')), 10_000);
      this.ws!.once('open', () => {
        clearTimeout(timer);
        resolve();
      });
      this.ws!.once('error', (err) => {
        clearTimeout(timer);
        reject(err);
      });
    });

    this.ws.on('message', (raw) => {
      let data: { transcript?: string; is_final?: boolean; confidence?: number };
      try {
        data = JSON.parse(raw.toString()) as typeof data;
      } catch {
        return;
      }

      const text = (data.transcript ?? '').trim();
      if (!text) return;

      const transcript: Transcript = {
        text,
        isFinal: Boolean(data.is_final),
        confidence: data.confidence ?? 0,
      };

      for (const cb of this.callbacks) {
        try {
          Promise.resolve(cb(transcript)).catch((err) =>
            getLogger().error(`STT transcript callback failed: ${String(err)}`),
          );
        } catch (err) {
          getLogger().error(`STT transcript callback threw: ${String(err)}`);
        }
      }
    });

    this.ws.on('error', (err) => {
      getLogger().warn(`TelnyxSTT WebSocket error: ${String(err)}`);
    });
  }

  /** Send a binary PCM16 audio chunk; emits the WAV header on the first call. */
  sendAudio(audio: Buffer): void {
    if (!this.ws || this.ws.readyState !== WebSocket.OPEN) return;

    if (!this.headerSent) {
      const header = createStreamingWavHeader(this.sampleRate, NUM_CHANNELS);
      this.ws.send(header);
      this.headerSent = true;
    }

    this.ws.send(audio);
  }

  /** Register a transcript listener. */
  onTranscript(callback: TranscriptCallback): void {
    this.callbacks.add(callback);
  }

  /** Unregister a previously-registered transcript listener. */
  offTranscript(callback: TranscriptCallback): void {
    this.callbacks.delete(callback);
  }

  /** Close the streaming WebSocket. */
  close(): void {
    if (this.ws) {
      try {
        this.ws.close();
      } catch {
        // ignore
      }
      this.ws = null;
    }
    this.headerSent = false;
  }
}

File: apps/patter-voice-sdk/libraries/python/getpatter/providers/telnyx_stt.py
"""Telnyx Speech-to-Text provider (WebSocket streaming).

Bridges the Telnyx ``/v2/speech-to-text/transcription`` WebSocket API to
Patter's :class:`~getpatter.providers.base.STTProvider` interface. Streams
PCM audio in and yields :class:`~getpatter.providers.base.Transcript`
events out via :meth:`receive_transcripts`.
"""

from __future__ import annotations

import asyncio
import json
import logging
import struct
from enum import IntEnum, StrEnum
from typing import ClassVar, AsyncIterator, Literal, Union
from urllib.parse import urlencode

import aiohttp

from getpatter.providers.base import STTProvider, Transcript

logger = logging.getLogger("getpatter.providers.telnyx_stt")

TELNYX_STT_WS_URL = "wss://api.telnyx.com/v2/speech-to-text/transcription"


class TelnyxTranscriptionEngine(StrEnum):
    """Backend transcription engines accepted by Telnyx STT."""

    TELNYX = "telnyx"
    GOOGLE = "google"
    DEEPGRAM = "deepgram"
    AZURE = "azure"


class TelnyxSTTSampleRate(IntEnum):
    """Common PCM sample rates accepted by Telnyx STT."""

    HZ_8000 = 8000
    HZ_16000 = 16000
    HZ_24000 = 24000


class TelnyxSTTInputFormat(StrEnum):
    """Input audio formats accepted by Telnyx STT."""

    WAV = "wav"


DEFAULT_SAMPLE_RATE = TelnyxSTTSampleRate.HZ_16000.value
NUM_CHANNELS = 1

# Backward-compatible Literal alias for type hints in older callers.
TranscriptionEngine = Literal["telnyx", "google", "deepgram", "azure"]


def _create_streaming_wav_header(sample_rate: int, num_channels: int) -> bytes:
    """Create a WAV header for streaming with maximum possible size."""
    bytes_per_sample = 2
    byte_rate = sample_rate * num_channels * bytes_per_sample
    block_align = num_channels * bytes_per_sample
    data_size = 0x7FFFFFFF
    file_size = 36 + data_size

    header = struct.pack(
        "<4sI4s4sIHHIIHH4sI",
        b"RIFF",
        file_size,
        b"WAVE",
        b"fmt ",
        16,
        1,
        num_channels,
        sample_rate,
        byte_rate,
        block_align,
        16,
        b"data",
        data_size,
    )
    return header


class TelnyxSTT(STTProvider):
    """Streaming STT adapter backed by Telnyx ``/v2/speech-to-text/transcription``.

    Args:
        api_key: Telnyx API key (Bearer token).
        language: Language code (e.g. ``"en"``, ``"es"``).
        transcription_engine: One of ``"telnyx"``, ``"google"``, ``"deepgram"``,
            ``"azure"``. Defaults to ``"telnyx"``.
        sample_rate: PCM sample rate in Hz. Defaults to 16 000.
        base_url: Override the base WebSocket URL (for testing).
        session: Optional pre-built ``aiohttp.ClientSession``. If omitted, a new
            session is created and closed with :meth:`close`.
    """

    #: Stable pricing/dashboard key — read by stream-handler/metrics.
    provider_key: ClassVar[str] = "telnyx_stt"

    def __init__(
        self,
        api_key: str,
        language: str = "en",
        *,
        transcription_engine: Union[
            TelnyxTranscriptionEngine, str
        ] = TelnyxTranscriptionEngine.TELNYX,
        sample_rate: Union[TelnyxSTTSampleRate, int] = TelnyxSTTSampleRate.HZ_16000,
        base_url: str = TELNYX_STT_WS_URL,
        session: aiohttp.ClientSession | None = None,
    ) -> None:
        self.api_key = api_key
        self.language = language
        self.transcription_engine = transcription_engine
        self.sample_rate = sample_rate
        self.base_url = base_url

        self._session = session
        self._owns_session = session is None
        self._ws: aiohttp.ClientWebSocketResponse | None = None
        self._header_sent = False
        self._queue: asyncio.Queue[Transcript | None] = asyncio.Queue()
        self._recv_task: asyncio.Task[None] | None = None
        # Bytes of audio forwarded to Telnyx since the last cost emission.
        # Used by ``_record_transcript_cost`` to compute ``patter.cost.stt_seconds``.
        self._audio_bytes_sent: int = 0

    def __repr__(self) -> str:
        return (
            f"TelnyxSTT(engine={self.transcription_engine!r}, "
            f"language={self.language!r}, sample_rate={self.sample_rate})"
        )

    def _record_transcript_cost(self) -> None:
        """Emit ``patter.cost.stt_seconds`` for audio buffered since the last
        final transcript. No-op when OTel is missing / no scope active."""
        try:
            from getpatter.observability.attributes import record_patter_attrs

            # Telnyx STT always receives 16-bit PCM (bytes_per_sample=2).
            seconds = self._audio_bytes_sent / float(self.sample_rate * 2)
            record_patter_attrs(
                {
                    "patter.cost.stt_seconds": seconds,
                    "patter.stt.provider": "telnyx",
                }
            )
            self._audio_bytes_sent = 0
        except Exception:  # pragma: no cover — defense in depth
            logger.debug("_record_transcript_cost failed", exc_info=True)

    async def connect(self) -> None:
        """Open the Telnyx Speech-to-Text WebSocket and start the recv loop."""
        if self._session is None:
            self._session = aiohttp.ClientSession()
            self._owns_session = True

        params = {
            "transcription_engine": self.transcription_engine,
            "language": self.language,
            "input_format": TelnyxSTTInputFormat.WAV.value,
        }
        url = f"{self.base_url}?{urlencode(params)}"
        headers = {"Authorization": f"Bearer {self.api_key}"}

        self._ws = await self._session.ws_connect(url, headers=headers)
        # Per-connection reset for sequential reuse: a second call on the
        # same instance never re-sent the WAV header (Telnyx then rejected
        # the raw PCM) and a stale ``None`` sentinel from the previous
        # close() instantly terminated the new transcript loop.
        self._header_sent = False
        while not self._queue.empty():
            try:
                self._queue.get_nowait()
            except asyncio.QueueEmpty:  # pragma: no cover - defensive
                break
        self._recv_task = asyncio.create_task(self._recv_loop())

    async def send_audio(self, audio_chunk: bytes) -> None:
        """Send a PCM audio chunk; prefixes the streaming WAV header on the first call."""
        if self._ws is None:
            raise RuntimeError("Not connected. Call connect() first.")

        if not self._header_sent:
            header = _create_streaming_wav_header(self.sample_rate, NUM_CHANNELS)
            await self._ws.send_bytes(header)
            self._header_sent = True

        await self._ws.send_bytes(audio_chunk)
        self._audio_bytes_sent += len(audio_chunk)

    async def receive_transcripts(self) -> AsyncIterator[Transcript]:
        """Yield :class:`Transcript` items as Telnyx returns transcription frames."""
        if self._ws is None:
            raise RuntimeError("Not connected. Call connect() first.")
        while True:
            item = await self._queue.get()
            if item is None:
                return
            yield item

    async def _recv_loop(self) -> None:
        assert self._ws is not None
        try:
            async for msg in self._ws:
                if msg.type == aiohttp.WSMsgType.TEXT:
                    parsed = self._parse_message(msg.data)
                    if parsed is not None:
                        if parsed.is_final:
                            self._record_transcript_cost()
                        await self._queue.put(parsed)
                elif msg.type in (
                    aiohttp.WSMsgType.CLOSE,
                    aiohttp.WSMsgType.CLOSED,
                    aiohttp.WSMsgType.CLOSING,
                ):
                    break
                elif msg.type == aiohttp.WSMsgType.ERROR:
                    break
        finally:
            await self._queue.put(None)

    @staticmethod
    def _parse_message(raw: str) -> Transcript | None:
        try:
            data = json.loads(raw)
        except (ValueError, TypeError):
            return None

        transcript = data.get("transcript", "")
        if not transcript:
            return None

        return Transcript(
            text=transcript,
            is_final=bool(data.get("is_final", False)),
            confidence=float(data.get("confidence", 0.0)),
        )

    async def close(self) -> None:
        """Cancel the recv task, close the socket, and release the HTTP session."""
        if self._recv_task is not None:
            self._recv_task.cancel()
            try:
                await self._recv_task
            except (asyncio.CancelledError, Exception):
                pass
            self._recv_task = None

        if self._ws is not None:
            try:
                await self._ws.close()
            except Exception:
                pass
            self._ws = None

        if self._owns_session and self._session is not None:
            try:
                await self._session.close()
            except Exception:
                pass
            self._session = None
